[FG:InPlacePodVerticalScaling] Remove restrictions on subresource flag in kubectl commands

AnishShah opened the pull request:
#### What type of PR is this?


/kind feature
/sig cli

#### What this PR does / why we need it:

We are introducing a new resize subresource in #128266. We would like to add support for this subresource in kubectl. We are doing this by removing restrictions on the `subresource` flag instead of adding `resize` to the allowlist.

#### Which issue(s) this PR fixes:

Fixes #128278

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?


```release-note
Removed restrictions on subresource flag in kubectl commands
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources
```

jpbetz commented on pkg/quota/v1/evaluator/core/resource_claims_test.go:
```diff
@@ -170,3 +172,52 @@ func TestResourceClaimEvaluatorMatchingResources(t *testing.T) {
 	}
 }
+
+func TestResourceClaimEvaluatorHandles(t *testing.T) {
```
Huge thanks for adding test coverage for these evaluators.

AnishShah committed:
kubectl: remove subresource restrictions from all commands

Removing this restrictions will allow us to use these commands with the
new resize subresource.

AnishShah commented:
/cc @ardaguclu @soltysh @liggitt

Initially I thought we would add `resize` subresource to the allowlist. I'm wondering if there's any issues with removing the restrictions on the subresource?

AnishShah commented:
/remove-area apiserver
/remove-area test
/triage accepted
/priority important-longterm
/remove-sig node
/remove-sig apps
/remove-sig api-machinery
/remove-sig architecture
/remove-sig insturmentation
/remove-sig testing

AnishShah commented:
/remove-sig instrumentation

liggitt commented:
> Initially I thought we would add resize subresource to the allowlist. I'm wondering if there's any issues with removing the restrictions on the subresource?

An allowlist we have to expand ad hoc hobbles old client binaries for no reason

liggitt commented:
lgtm, if we're removing the hard-coded limits on which subresources support these, should AddSubresourceFlags stop taking an allowedSubresources param?

ardaguclu commented:
Looks reasonable to me. I wonder was there any reason when this list was added. @soltysh do you recall?.

tallclair reviewed (commented):
With the removal of these restrictions, what happens if these commands are used on a subresource that takes a different schema? for example edit a `replicaset/scale`, or apply a `pod/bind`?

AnishShah commented:
> lgtm, if we're removing the hard-coded limits on which subresources support these, should AddSubresourceFlags stop taking an allowedSubresources param?

I removed this param.

> With the removal of these restrictions, what happens if these commands are used on a subresource that takes a different schema? for example edit a replicaset/scale, or apply a pod/bind?

I got this error:

```
kubectl patch pod pause -p '{"metadata":{"labels":{"env":"staging"}}}' --subresource=bind  
Error from server (NotFound): the server could not find the requested resource
```

liggitt commented:
> > With the removal of these restrictions, what happens if these commands are used on a subresource that takes a different schema? for example edit a replicaset/scale, or apply a pod/bind?
> 
> I got this error:
> 
> ```
> kubectl patch pod pause -p '{"metadata":{"labels":{"env":"staging"}}}' --subresource=bind  
> Error from server (NotFound): the server could not find the requested resource
> ```

Right, the operations just attempt to operate on the subresource specified... if the subresource doesn't exist or the submitted operation isn't valid for that subresource, it'll error.

tallclair commented:
/milestone v1.32

AnishShah commented:
kind ping @ardaguclu @soltysh since we are a few days away from code freeze.

#128266 is approved and ready to be merged. It would be good to get feedback on this PR so that the user can leverage new subresource via kubectl.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:
```diff
@@ -537,11 +537,8 @@ func AddPruningFlags(cmd *cobra.Command, prune *bool, pruneAllowlist *[]string,
 
-func AddSubresourceFlags(cmd *cobra.Command, subresource *string, usage string, allowedSubresources ...string) {
-	cmd.Flags().StringVar(subresource, "subresource", "", fmt.Sprintf("%s Must be one of %v. This flag is beta and may change in the future.", usage, allowedSubresources))
-	CheckErr(cmd.RegisterFlagCompletionFunc("subresource", func(*cobra.Command, []string, string) ([]string, cobra.ShellCompDirective) {
```
It makes sense to remove the validations but why are we losing completions?. I think, we can still pass the list (by also adding `resize`) to this function.

AnishShah replied:
Done.

ardaguclu commented:
Thank you
/lgtm
/approve
